src/coprocessor/statistics/histogram.rs

```diff
@@ -0,0 +1,235 @@
+        bucket.set_repeats(self.repeats);
+        bucket.set_count(self.count);
+        bucket.set_lower_bound(self.lower_bound);
+        bucket.set_upper_bound(self.upper_bound);
+        bucket
+    }
+}
+
+/// Histogram represents statistics for a column or index.
+#[derive(Default)]
+pub struct Histogram {
+    // column's id
+    id: i64,
+    // number of distinct values
+    ndv: i64,
+    buckets: Vec<Bucket>,
+    // max number of values in per bucket. Notice: when a bucket's count is equal to
+    // per_bucket_limit, only value equal with last value can been inserted into it.
+    per_bucket_limit: i64,
+    // max number of buckets
+    buckets_num: usize,
+}
+
+impl Histogram {
+    pub fn new(id: i64, buckets_num: usize) -> Histogram {
+        let mut h = Histogram::default();
+        h.per_bucket_limit = 1;
+        h.buckets_num = buckets_num;
+        h.id = id;
+        h
+    }
+
+    pub fn into_proto(self) -> analyze::Histogram {
+        let mut hist = analyze::Histogram::new();
+        hist.set_ndv(self.ndv);
+        let buckets: Vec<analyze::Bucket> = self.buckets
+            .into_iter()
+            .map(|bucket| bucket.into_proto())
+            .collect();
+        hist.set_buckets(RepeatedField::from_vec(buckets));
+        hist
+    }
+
+    pub fn iterate(&mut self, data: Vec<u8>) {
+        if let Some(bucket) = self.buckets.last_mut() {
+            // The new item has the same value as last bucket value, to ensure that
+            // a same value only stored in a single bucket, we do not increase bucket
+            // even if it exceeds per_bucket_limit.
+            if bucket.upper_bound == data {
+                bucket.insert_repeated_item();
+                return;
+            }
+        }
+        self.ndv += 1;
+        if self.buckets.len() >= self.buckets_num && self.is_last_bucket_full() {
+            self.merge_buckets();
+        }
+
+        if !self.is_last_bucket_full() {
+            self.buckets.last_mut().unwrap().insert_item(data);
+            return;
+        }
+
+        // create a new bucket and insert data
+        let mut count = 1;
+        if let Some(bucket) = self.buckets.last() {
+            count += bucket.count
+        }
+        self.buckets.push(Bucket::new(count, data.clone(), data, 1));
+    }
+
+    // check whether the last bucket is full.
+    fn is_last_bucket_full(&self) -> bool {
+        if self.buckets.is_empty() {
+            return true;
+        }
+        let count = if self.buckets.len() == 1 {
+            self.buckets[0].count
+        } else {
+            let len = self.buckets.len();
+            self.buckets[len - 1].count - self.buckets[len - 2].count
+        };
+        count >= self.per_bucket_limit
+    }
+
+    // It merges every two neighbor buckets.
+    fn merge_buckets(&mut self) {
+        self.buckets = {
+            let mut buckets = Vec::with_capacity(self.buckets.len() / 2 + (self.buckets.len() & 1));
```

Comment:
Is it necessary if buckets is going to reach `bucket_num`? What's the maximum possible value of `bucket_num`?

Reply:
@lamxTyler

Reply:
Only necessary when we are going to append a value and all buckets are full. Currently, the max value of `bucket_num` is 256.

Reply:
Then it's unnecessary to reallocate a new vector every merge. And you can also use the for loop to edit buckets in place.

```
for i in 0..(self.buckets_num + 1) / 2 {
    self.buckets[i] = calculation on (self.buckets[i * 2], self.buckets[i * 2 + 1])
}
```